### PR TITLE
Fixes to Serialization

### PR DESCRIPTION
**Pull request by atzannes**
* Additional serializaiton tests
* Support for `set()` and `frozenset()`
* Refactorings

### CODE REVIEW

File: typed_python/PythonSerializationContext.hpp [hunk 1]
```diff
@@ -153,89 +236,61 @@ class PythonSerializationContext : public SerializationContext {
...
+        if (!set) {
+            return set;
+        }
 
+        PyObject* res = nullptr;
         try {
-            for (long k = 0; k < sz; k++) {
-                PyTuple_SET_ITEM(res, k, deserializePythonObject(b));
-            }
+            res = PyFrozenSet_New(PyObject_GetIter(set));
```
Comment (braxtonmckee) [inline typed_python/PythonSerializationContext.hpp]: don't forget to decref the iterator, and also to check for exceptions during set construction

File: typed_python/PythonSerializationContext.hpp [hunk 2]
```diff
@@ -655,6 +713,105 @@ class PythonSerializationContext : public SerializationContext {
...
+        }
+
+        size_t sz = b.read_uint32();
+
+        res = factory_fn(NULL);
+        b.addCachedPointer(id, incref(res), true);
+
+        try {
+            for (long k = 0; k < sz; k++) {
+                int success = add_item_fn(res, deserializePythonObject(b));
```
Comment (braxtonmckee) [inline typed_python/PythonSerializationContext.hpp]: are we decreffing the result of 'deserializePythonObject' anywhere?

Comment (atzannes) [reply typed_python/PythonSerializationContext.hpp]: Good question! So `add_item_fn` is `PySet_Add`. It's documentation does not specify if it steals the reference to the key being added to the set. Theoretically it should to keep symmetry with `PyList_SetItem` and `PyTuple_SetItem`.

Comment (atzannes) [reply typed_python/PythonSerializationContext.hpp]: Baaaah!! Looking at the code it looks like `PySet_Add` does not steal the reference. It does `Py_INCREF(key)`. Bogus!

File: typed_python/PythonSerializationContext.hpp [hunk 3]
```diff
@@ -655,6 +739,113 @@ class PythonSerializationContext : public SerializationContext {
...
+            throw std::runtime_error(std::string(
+                "Failed to allocate storage into which to deserialize iterable"));
+        }
+        b.addCachedPointer(id, incref(res), true);
+
+        try {
+            for (long k = 0; k < sz; k++) {
+                PyObject* item = deserializePythonObject(b);
+                int success = add_item_fn(res, item);
+                Py_DECREF(item);
```
Comment (atzannes) [inline typed_python/PythonSerializationContext.hpp]: ok, making sure to decref the item after adding it to a set

File: typed_python/PythonSerializationContext.hpp [hunk 4]
```diff
@@ -216,14 +272,38 @@ class PythonSerializationContext : public SerializationContext {
...
+                // currently being deserialized. In that case PySet_Add will fail, so we temporarily
+                // decref it.
+                auto refcount = Py_REFCNT(res);
+                for (int i = 1; i < refcount; i++) Py_DECREF(res);
+
+                int success = PySet_Add(res, item);
+
+                for (int i = 1; i < refcount; i++) Py_INCREF(res);
+
+                Py_DECREF(item);
```
Comment (atzannes) [inline typed_python/PythonSerializationContext.hpp]: `PySet_Add` does not steal the reference, so we must decref it

**Comment by atzannes** [comment]
@braxtonmckee this is ready for another look from you. I added a number of serialization tests that are failing (which I marked with `@unittest.skip`) so that we have an idea of some of the things that pickle does that we're not, but not have to implement them right-away.

**Comment by atzannes** [comment]
@braxtonmckee this is the second PR. I've rebased it on top of #16

**Pull request merged**